base/reduce.jl

```diff
@@ -1246,3 +1184,64 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
     end
     return n
 end
+
+# simd optimization for min/max related reduction
+# general fallback
+_fast(op, x, y) = op(x, y)
+_fast(op) = (x, y) -> _fast(op, x, y)
+
+# The following are used in optimized `mapreduce_impl` for IEEEFloat.
+_fast(::typeof(min), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(x-y), x, y)
+_fast(::typeof(max), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(y-x), x, y)
+
+# Some help function for better vectorization
+# NaN check
+_anynan(x, y) = isnan(x) | isnan(y)
+_anynan(x::NTuple{N}) where {N} = reduce(|, map(_anynan, x[1:N>>1], x[1+N>>1:N]))
+# split then reduce for inputs whose length > 4
+_half_reduce(op, x::NTuple{N}) where {N} = _half_reduce(op, map(op, x[1:N>>1], x[1+N>>1:N]))
+_half_reduce(op, x::NTuple{4}) = reduce(op, x)
+
+function mapreduce_impl(f, op::Union{typeof(max),typeof(min)},
+                        A::AbstractArrayOrBroadcasted, first::Int, last::Int)
+    T = @default_eltype A
+    Eltype = _return_type(f, Tuple{T})
+    # Limit this optimization to concrete IEEEFloat. (general fallback is 2x faster for Integer)
+    (isconcretetype(Eltype) && Eltype <: IEEEFloat) ||
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
```

Comment:
We need to leave some comments in the source code to explain why this is an optimization; i.e., it makes the function faster without changing any observable behaviors.

Actually, I take back some of my complaints in https://github.com/JuliaLang/julia/pull/43604#discussion_r780725734. Since we are restricting the following code to `IEEEFloat`, nan is the only "degenerate" element that contributes to the non-commutativity of `min` and `max`. However, we are making sure that the first occurrence of nan is returned. As such, the result is `===`-equivalent with and without this method specialization.

Does this sound correct? If so, please explain the idea in the source code.

Reply:
That’s correct, l’ll add some comments.
In factor, if we use general fallback, the first nan might not returned, as Base.min thought copysign(NaN, -1) < NaN.

Reply:
For reference, LLVM return first `NaN`. (https://reviews.llvm.org/D52764)
Standard:
 > If two or more inputs are NaN, then the payload of the resulting NaN should be identical to the payload of
one of the input NaNs if representable in the destination format. This standard does not specify which of
the input NaNs will provide the payload.

Reply:
Since our `min/max` definition, ensure `===` with general fallback seems non-sense at current stage.
For example, if `A[1] = NaN`, `minimum` needs to search the rest of `A` to check is there a `copysign(NaN, -1)`.

Reply:
I left `min`/`max` as `TODO`. Comments added.